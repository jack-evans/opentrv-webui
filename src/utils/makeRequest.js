
export default function makeRequest (url, options) {
  return global.fetch(url, options)
    .then(internal.checkStatus)
    .then(internal.parseJSON)
}

function checkStatus (response) {
  if (response.status >= 200 && response.status < 300) {
    return response
  } else {
    let error = new Error(response.statusText)
    error.response = response
    throw error
  }
}

function parseJSON (response) {
  return response.json()
}

export const internal = {
  checkStatus: checkStatus,
  parseJSON: parseJSON
}
