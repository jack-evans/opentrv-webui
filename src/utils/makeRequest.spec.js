import * as makeRequest from './makeRequest'

describe('makeRequest.js', () => {
  const fakeUrl = 'http://localhost:3002'
  let fetchSpy

  beforeEach(() => {
    fetchSpy = jest.spyOn(global, 'fetch')
  })

  afterEach(() => {
    fetchSpy.mockReset()
  })

  describe('default exported function', () => {
    it('calls the checkStatus function', () => {
      fetchSpy.mockReturnValue(Promise.resolve())
      const checkStatusSpy = jest.spyOn(makeRequest.internal, 'checkStatus').mockReturnValue(Promise.resolve())
      const parseJSONSpy = jest.spyOn(makeRequest.internal, 'parseJSON').mockReturnValue(Promise.resolve())

      return makeRequest.default(fakeUrl + '/examplePath', {method: 'POST'})
        .then(() => {
          expect(checkStatusSpy).toHaveBeenCalledTimes(1)
          checkStatusSpy.mockRestore()
          parseJSONSpy.mockRestore()
        })
    })

    it('calls the parseJSON function', () => {
      fetchSpy.mockReturnValue(Promise.resolve())
      const checkStatusSpy = jest.spyOn(makeRequest.internal, 'checkStatus').mockReturnValue(Promise.resolve())
      const parseJSONSpy = jest.spyOn(makeRequest.internal, 'parseJSON').mockReturnValue(Promise.resolve())

      return makeRequest.default(fakeUrl + '/examplePath', {method: 'POST'})
        .then(() => {
          expect(parseJSONSpy).toHaveBeenCalledTimes(1)
          checkStatusSpy.mockRestore()
          parseJSONSpy.mockRestore()
        })
    })
  })

  describe('internal functions', () => {
    describe('checkStatus', () => {
      describe('when the response status is between 200 and 300', () => {
        it('returns the response', () => {
          const response = {
            status: 201,
            statusText: 'OK',
            body: {
              somecontent: 'content'
            }
          }

          expect(makeRequest.internal.checkStatus(response)).toEqual(response)
        })
      })

      describe('when the response is not between 200 and 300', () => {
        it('throws an error', () => {
          const response = {
            status: 500,
            statusText: 'Internal server error'
          }

          expect(() => {
            makeRequest.internal.checkStatus(response)
          }).toThrowErrorMatchingSnapshot()
        })
      })
    })

    describe('parseJSON', () => {
      it('calls the json method within the response', () => {
        const response = {
          status: 200,
          statusText: 'OK',
          json: () => {}
        }

        const jsonSpy = jest.spyOn(response, 'json')

        makeRequest.internal.parseJSON(response)
        expect(jsonSpy).toHaveBeenCalledTimes(1)
        jsonSpy.mockReset()
      })
    })
  })
})
