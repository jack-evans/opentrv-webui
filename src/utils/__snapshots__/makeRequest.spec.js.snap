// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeRequest.js internal functions checkStatus when the response is not between 200 and 300 throws an error 1`] = `"Internal server error"`;
